pkg/controller/instance/plan_execution_engine.go

```diff
@@ -1,332 +1,228 @@
 package instance
 
 import (
-	"context"
-	"encoding/json"
+	"errors"
 	"fmt"
 	"log"
-	"strconv"
-
-	"k8s.io/apimachinery/pkg/types"
 
 	"github.com/kudobuilder/kudo/pkg/apis/kudo/v1alpha1"
-	kudoengine "github.com/kudobuilder/kudo/pkg/engine"
-	"github.com/kudobuilder/kudo/pkg/util/health"
-	errwrap "github.com/pkg/errors"
-	apiextv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
-	apierrors "k8s.io/apimachinery/pkg/api/errors"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/runtime"
-	apijson "k8s.io/apimachinery/pkg/util/json"
+	engtask "github.com/kudobuilder/kudo/pkg/engine/task"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+var (
+	unknownTaskNameEventName         = "UnknownTaskName"
+	unknownTaskKindEventName         = "UnknownTaskName"
+	fatalTaskExecutionErrorEventName = "FatalTaskExecutionError"
+)
+
 type activePlan struct {
 	name string
 	*v1alpha1.PlanStatus
 	spec      *v1alpha1.Plan
-	tasks     map[string]v1alpha1.TaskSpec
+	tasks     []v1alpha1.Task
 	templates map[string]string
 	params    map[string]string
 }
 
-type planResources struct {
-	PhaseResources map[string]phaseResources
-}
-
-type phaseResources struct {
-	StepResources map[string][]runtime.Object
-}
-
-type engineMetadata struct {
-	instanceName        string
-	instanceNamespace   string
-	operatorName        string
-	operatorVersionName string
-	operatorVersion     string
-
-	// the object that will own all the resources created by this execution
-	resourcesOwner metav1.Object
+func (ap *activePlan) taskByName(name string) (*v1alpha1.Task, bool) {
```

Comment:
does it also make you sad that you have to write functions like this by hand? 😂

Reply:
Every time! In Scala that would've been: `ap.tasks.find(_.Name == name)` 😞